File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/jsx-no-comment-textnodes.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";

const MESSAGE =
  "Your users see this comment as text on the page because `//` & `/*` aren't hidden in JSX.";

// HTML elements that intentionally render their text content
// verbatim — `<code>//# chunkId=</code>`, `<pre>// build output</pre>`,
// `<kbd>Ctrl/Cmd</kbd>`, `<samp>// no errors</samp>`. Text inside
// these tags is meant to be literal; flagging `//` in their bodies
// is a guaranteed false positive.
const LITERAL_TEXT_TAGS: ReadonlySet<string> = new Set(["code", "pre", "kbd", "samp", "tt"]);

const hasCommentLikePattern = (text: string, followsExpressionContainer: boolean): boolean => {
  for (const [lineIndex, rawLine] of text.split("\n").entries()) {
    const trimmed = rawLine.trim();
    if (trimmed.startsWith("/*")) return true;
    if (!trimmed.startsWith("//")) continue;
    // A line that trims to just `//` (or `// `) is an interpolated
    // separator glyph (`{used} // {total} GB`), not a `// comment` —
    // require some comment body after the slashes before flagging.
    if (trimmed.slice(2).trim().length === 0) continue;
    // The same separator idiom with a literal numeric right side
    // (`{used} // 512 GB`) — a first line that continues a preceding
    // `{expression}` with a digit-leading value is deliberate rendered
    // text, not a stray comment. Prose after the slashes
    // (`{value} // visible to users`) is still a comment; later lines
    // start fresh and are still checked.
    if (lineIndex === 0 && followsExpressionContainer && /^\d/.test(trimmed.slice(2).trimStart()))
      continue;
    return true;
  }
  return false;
};

const followsExpressionContainerSibling = (node: EsTreeNodeOfType<"JSXText">): boolean => {
  const parent = node.parent;
  if (!parent || (!isNodeOfType(parent, "JSXElement") && !isNodeOfType(parent, "JSXFragment"))) {
    return false;
  }
  const siblingIndex = parent.children.findIndex((child) => child === node);
  if (siblingIndex <= 0) return false;
  const previousSibling = parent.children[siblingIndex - 1];
  return (
    isNodeOfType(previousSibling, "JSXExpressionContainer") &&
    !isNodeOfType(previousSibling.expression, "JSXEmptyExpression")
  );
};

const isInsideLiteralTextTag = (node: EsTreeNode): boolean => {
  let ancestor: EsTreeNode | null | undefined = node.parent;
  while (ancestor) {
    if (isNodeOfType(ancestor, "JSXElement")) {
      const openingName = ancestor.openingElement.name as EsTreeNode;
      if (isNodeOfType(openingName, "JSXIdentifier") && LITERAL_TEXT_TAGS.has(openingName.name)) {
        return true;
      }
    }
    ancestor = ancestor.parent ?? null;
  }
  return false;
};

// Port of `oxc_linter::rules::react::jsx_no_comment_textnodes`. Reports
// JSX text nodes whose lines start with `//` or `/*` — these aren't
// comments, they're rendered as literal text. Skips text inside
// `<code>`, `<pre>`, `<kbd>`, `<samp>`, `<tt>` — those tags exist to
// render literal text including code-comment-like prefixes.
export const jsxNoCommentTextnodes = defineRule({
  id: "jsx-no-comment-textnodes",
  title: "Comment rendered as JSX text",
  severity: "warn",
  recommendation:
    "Wrap JSX comments in `{/* … */}` so users do not see comment text rendered as children.",
  create: (context) => ({
    JSXText(node: EsTreeNodeOfType<"JSXText">) {
      if (!hasCommentLikePattern(node.value, followsExpressionContainerSibling(node))) return;
      if (isInsideLiteralTextTag(node)) return;
      context.report({ node, message: MESSAGE });
    },
  }),
});

File: native/oxlint/rules/jsx-no-comment-textnodes.rs
use oxc_ast::{
    AstKind,
    ast::{JSXAttributeItem, JSXAttributeName, JSXChild, JSXElementName, JSXExpression},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
};

const MESSAGE: &str =
    "Your users see this comment as text on the page because `//` & `/*` aren't hidden in JSX.";
const LITERAL_TEXT_TAGS: [&str; 5] = ["code", "pre", "kbd", "samp", "tt"];
const STYLING_ATTRIBUTE_NAMES: [&str; 3] = ["className", "class", "style"];

#[derive(Debug, Default, Clone)]
pub struct JsxNoCommentTextnodes;

declare_oxc_lint!(
    /// Disallow accidental comment text rendered as JSX children.
    JsxNoCommentTextnodes,
    react_doctor_native,
    suspicious,
    version = "0.1.0",
    short_description = "Disallow accidental comment text rendered as JSX children.",
);

impl Rule for JsxNoCommentTextnodes {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::JSXText(jsx_text) = node.kind() else {
            return;
        };
        if !has_comment_like_pattern(&jsx_text.value, follows_expression_container(node, ctx))
            || is_inside_literal_text_tag(node, ctx)
            || is_deliberate_styled_comment_token(node, ctx)
        {
            return;
        }
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(jsx_text.span));
    }

    fn should_run(&self, ctx: &ContextHost) -> bool {
        ctx.source_type().is_jsx()
    }
}

fn has_comment_like_pattern(text: &str, follows_expression_container: bool) -> bool {
    text.split('\n').enumerate().any(|(line_index, raw_line)| {
        let trimmed = raw_line.trim();
        if trimmed.starts_with("/*") {
            return true;
        }
        let Some(comment_body) = trimmed.strip_prefix("//") else {
            return false;
        };
        let trimmed_comment_body = comment_body.trim_start();
        if trimmed_comment_body.is_empty() {
            return false;
        }
        !(line_index == 0
            && follows_expression_container
            && trimmed_comment_body
                .as_bytes()
                .first()
                .is_some_and(u8::is_ascii_digit))
    })
}

fn follows_expression_container(node: &AstNode, ctx: &LintContext) -> bool {
    let parent = ctx.nodes().parent_node(node.id());
    let children = match parent.kind() {
        AstKind::JSXElement(element) => &element.children,
        AstKind::JSXFragment(fragment) => &fragment.children,
        _ => return false,
    };
    let Some(child_index) = children
        .iter()
        .position(|child| child.span() == node.span())
    else {
        return false;
    };
    if child_index == 0 {
        return false;
    }
    matches!(
        &children[child_index - 1],
        JSXChild::ExpressionContainer(container)
            if !matches!(container.expression, JSXExpression::EmptyExpression(_))
    )
}

fn is_inside_literal_text_tag(node: &AstNode, ctx: &LintContext) -> bool {
    ctx.nodes().ancestors(node.id()).any(|ancestor| {
        let AstKind::JSXElement(element) = ancestor.kind() else {
            return false;
        };
        matches!(
            &element.opening_element.name,
            JSXElementName::Identifier(identifier)
                if LITERAL_TEXT_TAGS.contains(&identifier.name.as_str())
        )
    })
}

fn is_deliberate_styled_comment_token(node: &AstNode, ctx: &LintContext) -> bool {
    let AstKind::JSXText(jsx_text) = node.kind() else {
        return false;
    };
    let trimmed = jsx_text.value.trim();
    if !trimmed.starts_with("//") && !trimmed.starts_with("/*") {
        return false;
    }
    let parent = ctx.nodes().parent_node(node.id());
    let AstKind::JSXElement(element) = parent.kind() else {
        return false;
    };
    if element.children.iter().any(|child| match child {
        JSXChild::Text(text) => text.span != jsx_text.span && !text.value.trim().is_empty(),
        _ => child.span() != jsx_text.span,
    }) {
        return false;
    }
    element.opening_element.attributes.iter().any(|attribute| {
        let JSXAttributeItem::Attribute(attribute) = attribute else {
            return false;
        };
        matches!(
            &attribute.name,
            JSXAttributeName::Identifier(identifier)
                if STYLING_ATTRIBUTE_NAMES.contains(&identifier.name.as_str())
        )
    })
}
